Mathlib/Algebra/Star/Module.lean

```diff
@@ -0,0 +1,168 @@
+    [Module R A] [StarModule R A] : A ≃ₗ⋆[R] A :=
+  { starAddEquiv with
+    toFun := star
+    map_smul' := star_smul }
+#align star_linear_equiv starLinearEquiv
+
+end deinstance
+
+variable (R : Type _) (A : Type _) [Semiring R] [StarSemigroup R] [TrivialStar R] [AddCommGroup A]
+  [Module R A] [StarAddMonoid A] [StarModule R A]
+
+/-- The self-adjoint elements of a star module, as a submodule. -/
+def selfAdjoint.submodule : Submodule R A :=
+  { selfAdjoint A with smul_mem' := IsSelfAdjoint.smul }
+#align self_adjoint.submodule selfAdjoint.submodule
+
+/-- The skew-adjoint elements of a star module, as a submodule. -/
+def skewAdjoint.submodule : Submodule R A :=
+  { skewAdjoint A with smul_mem' := skewAdjoint.smul_mem }
+#align skew_adjoint.submodule skewAdjoint.submodule
+
+variable {A} [Invertible (2 : R)]
+
+/-- The self-adjoint part of an element of a star module, as a linear map. -/
+@[simps]
+def selfAdjointPart : A →ₗ[R] selfAdjoint A
+    where
+  toFun x :=
+    ⟨(⅟ 2 : R) • (x + star x), by
+      simp only [selfAdjoint.mem_iff, star_smul, add_comm, StarAddMonoid.star_add, star_inv',
+        star_bit0, star_one, star_star, star_invOf (2 : R), star_trivial]⟩
+  map_add' x y := by
+    ext
+    simp [add_add_add_comm]
+  map_smul' r x := by
+    ext
+    simp [← mul_smul, show ⅟ 2 * r = r * ⅟ 2 from Commute.invOf_left <| (2 : ℕ).cast_commute r]
+#align self_adjoint_part selfAdjointPart
+
+/-- The skew-adjoint part of an element of a star module, as a linear map. -/
+@[simps]
+def skewAdjointPart : A →ₗ[R] skewAdjoint A
+    where
+  toFun x :=
+    ⟨(⅟ 2 : R) • (x - star x), by
+      simp only [skewAdjoint.mem_iff, star_smul, star_sub, star_star, star_trivial, ← smul_neg,
+        neg_sub]⟩
+  map_add' x y := by
+    ext
+    simp only [sub_add, ← smul_add, sub_sub_eq_add_sub, star_add, AddSubgroup.coe_mk,
+      AddSubgroup.coe_add]
+  map_smul' r x := by
+    ext
+    simp [← mul_smul, ← smul_sub,
+      show r * ⅟ 2 = ⅟ 2 * r from Commute.invOf_right <| (2 : ℕ).commute_cast r]
+#align skew_adjoint_part skewAdjointPart
+
+theorem StarModule.selfAdjointPart_add_skewAdjointPart (x : A) :
+    (selfAdjointPart R x : A) + skewAdjointPart R x = x := by
+  simp only [smul_sub, selfAdjointPart_apply_coe, smul_add, skewAdjointPart_apply_coe,
+    add_add_sub_cancel, inv_of_two_smul_add_inv_of_two_smul]
+#align star_module.self_adjoint_part_add_skew_adjoint_part StarModule.selfAdjointPart_add_skewAdjointPart
+
+variable (A)
+
+/-- The decomposition of elements of a star module into their self- and skew-adjoint parts,
+as a linear equivalence. -/
+@[simps!]
+def StarModule.decomposeProdAdjoint : A ≃ₗ[R] selfAdjoint A × skewAdjoint A :=
+  LinearEquiv.ofLinear ((selfAdjointPart R).prod (skewAdjointPart R))
+    ((selfAdjoint.submodule R A).subtype.coprod (skewAdjoint.submodule R A).subtype)
+    (by ext <;> simp <;> sorry) (LinearMap.ext <| StarModule.selfAdjointPart_add_skewAdjointPart R)
```

Comment:
I had a look at this and was able to make progress of a sort but things became quite ugly.
```suggestion
    (by
      refine LinearMap.prod_ext (LinearMap.ext $ fun x ↦ Prod.ext (Subtype.ext ?_) (Subtype.ext ?_))
                                (LinearMap.ext $ fun x ↦ Prod.ext (Subtype.ext ?_) (Subtype.ext ?_))
      · simp only [LinearMap.coe_comp, Function.comp_apply, LinearMap.prod_apply]
        erw [LinearMap.coprod_apply] -- Works but `simp only`, `rw`, `simp_rw` all fail
        simp
      · sorry
      · sorry
      · sorry)
      (LinearMap.ext <| StarModule.selfAdjointPart_add_skewAdjointPart R)
```
For me the takeaways are:

- The `ext` tactic is failing to apply all the extensionality lemmas it should even though they work if applied manually
- The lemma `LinearMap.coprod_apply` gets singled out: for some reason `simp` and `rw` are unwilling to apply it but `erw` will work
- This is far too ugly: I presume the thing to do is to wait for https://github.com/leanprover/lean4/issues/2074 to be resolved (which I hope fixes this)

Reply:
Are the `ext` lemmas not being applied because they don't have a high enough priority?

Reply:
I just had a look at it two and I think I have a compiling version :D (lean4#2074 did not make a difference). Maybe that helps finding the issues underlying it.

Reply:
@j-loreaux I didn't know that extensionality lemmas could carry priority. Can you show me an example of this?

@joneugster Great work closing the goal, and especially demonstrating that this is not completely solved by https://github.com/leanprover/lean4/issues/2074 I think we need to do some more work to understand the various issues at play here though.

I hope to look into this again over the weekend unless someone else gets there before me.

Reply:
@ocfnash have a look at [QuotientGroup.monoidHom_ext](https://github.com/leanprover-community/mathlib4/blob/17abfa705cdef525a0d7e3c41d3b78c1fed8c9aa/Mathlib/GroupTheory/QuotientGroup.lean#L117-L118). The issue is that this is an extensionality for `monoidHom`s that is supposed to apply only when the group is a quotient group. In Lean 3, this would have been tried first because it appears later than the usual `MonoidHom.ext`, but in Lean 4 it needs a priority setting in order to guarantee that it is tried first. So basically, you should assign priorities whenever you have two `ext` lemmas that apply to the same underlying type (in this case, `monoidHom`).